runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10378,45 +10364,113 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
+                  }
+               }
 
-         iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, shiftReg, fillerLoopLabel);
-         iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerRemLabel);
+            // Zero blocks of 256 bytes if size > 256.
+            iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BZ, node, fillerRemLabel, iCursor);
+            iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerLoopLabel);
+            iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 255, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+            iCursor = generateRXInstruction(cg, TR::InstOpCode::LA, node, addressReg, generateS390MemoryReference(addressReg, 256, cg), iCursor);
+            iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, shiftReg, fillerLoopLabel);
+            iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, fillerRemLabel);
+
+            // Zero the remainder.
+            TR::LabelSymbol *exrlTargetLabel = generateLabelSymbol(cg);
+            iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, exrlTargetLabel);
+            iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 0, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+            iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerRemLabel);
+            iCursor = generateRILInstruction(cg, TR::InstOpCode::EXRL, node, lengthReg, exrlTargetLabel);
+            }
+         else
+            {
+            
+            int32_t remainder = allocSize % 256;
+            int32_t displacement = 0;
+            // Using a BRCT loop to clean blocks of 256 to save iCache.
+            if (useBrtc)
+               {
+               shiftReg = cg->allocateRegister();
+               if (conditions != NULL)
+                  conditions->addPostCondition(shiftReg, TR::RealRegister::AssignAny);
+
+               iCursor = generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, addressReg, resReg, iCursor);
+               iCursor = generateLoad32BitConstant(cg, node, loops, shiftReg, true, iCursor, conditions);
+               iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerLoopLabel);
+               iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 255, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+               iCursor = generateRXInstruction(cg, TR::InstOpCode::LA, node, addressReg, generateS390MemoryReference(addressReg, 256, cg), iCursor);
+               iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, shiftReg, fillerLoopLabel);
+               loops = 0;
```

Comment:
I understand what you are doing here, but isn't this equivalent to, Also `useBrtc` is really odd name and some one who has written this code or been reviewing (Effectively you and me) can understand purpose of it. Please use some better naming for that (i.e , `generateXCLoop`
```
if (useBrtc)
...
else
...
```

Reply:
even with this else condition, i still need the `while (loops > 0)` while so I thought it might be shorter to set loops to zero instead of adding an else. I'll change it to make it more readable.

Reply:
> even with this else condition, i still need the while (loops > 0) while so I thought it might be shorter to set loops to zero instead of adding an else.

How it is shorter? If I understand correctly, If you are generating XC loop, in which case you do not need to generate constant length case, you will still generate a loop which will not entered. In current case, you will always generate loop and will always check if you need to enter in to loop or not even in cases when you will not.

Reply:
By `shorter` I meant less lines of code not shorter for execution. I removed the `loops = 0` from the if condition and put the `while` in the else as you suggested.